oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSuggestIterator.java

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.index.elastic.query;
+
+import org.apache.jackrabbit.oak.plugins.index.search.FieldNames;
+import org.apache.jackrabbit.oak.plugins.index.search.spi.query.FulltextIndex.FulltextResultRow;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.SortBuilders;
+import org.elasticsearch.search.sort.SortOrder;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * This class is in charge to extract suggestions for a given query. Suggestion is more like
+ * a completion result.
+ */
+class ElasticSuggestIterator implements Iterator<FulltextResultRow> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticSuggestIterator.class);
+    protected final static String SUGGEST_PREFIX = "suggest?term=";
+
+    private final ElasticIndexNode indexNode;
+    private final ElasticRequestHandler requestHandler;
+    private final ElasticResponseHandler responseHandler;
+    private final String suggestQuery;
+
+    private Iterator<FulltextResultRow> internalIterator;
+    private boolean loaded = false;
+
+    ElasticSuggestIterator(@NotNull ElasticIndexNode indexNode,
+                           @NotNull ElasticRequestHandler requestHandler,
+                           @NotNull ElasticResponseHandler responseHandler) {
+        this.indexNode = indexNode;
+        this.requestHandler = requestHandler;
+        this.responseHandler = responseHandler;
+        this.suggestQuery = requestHandler.getPropertyRestrictionQuery().replace(SUGGEST_PREFIX, "");
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!loaded) {
+            try {
+                loadSuggestions();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            loaded = true;
+        }
+        return internalIterator != null && internalIterator.hasNext();
+    }
+
+    @Override
+    public FulltextResultRow next() {
+        return internalIterator.next();
+    }
+
+    private void loadSuggestions() throws IOException {
+        BoolQueryBuilder suggestionQuery = requestHandler.suggestionMatchQuery(suggestQuery);
+        final SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource()
+                .query(suggestionQuery)
+                .size(100)
+                .fetchSource(FieldNames.PATH, null);
+        final SearchRequest searchRequest = new SearchRequest(indexNode.getDefinition().getRemoteIndexAlias())
```

Comment:
I wouldn't use "final" here, because it clutters the code a bit, and is not needed.